Jetsnack/app/src/main/java/com/example/jetsnack/ui/JetsnackApp.kt

```diff
@@ -17,30 +17,155 @@
 import com.example.jetsnack.ui.home.JetsnackBottomBar
 import com.example.jetsnack.ui.theme.JetsnackTheme
 import com.google.accompanist.insets.ProvideWindowInsets
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.receiveAsFlow
+import kotlinx.coroutines.launch
 
 @Composable
 fun JetsnackApp() {
     ProvideWindowInsets {
         JetsnackTheme {
-            val tabs = remember { HomeSections.values() }
-            val navController = rememberNavController()
+            val scaffoldState = rememberScaffoldState()
+            val scaffoldStateHolder = rememberScaffoldStateHolder(scaffoldState)
+
             JetsnackScaffold(
-                bottomBar = { JetsnackBottomBar(navController = navController, tabs = tabs) }
+                scaffoldState = scaffoldState,
+                bottomBar = {
+                    JetsnackBottomBar(scaffoldStateHolder.navController, scaffoldStateHolder.tabs)
+                }
             ) { innerPaddingModifier ->
                 JetsnackNavGraph(
-                    navController = navController,
+                    scaffoldStateHolder = scaffoldStateHolder,
                     modifier = Modifier.padding(innerPaddingModifier)
                 )
             }
         }
     }
 }
+
+/**
+ * Creates a [ScaffoldStateHolder] and memoizes it.
+ *
+ * Pending snackbar messages to show on the screen are remembered across
+ * activity and process recreation.
+ */
+@Composable
+private fun rememberScaffoldStateHolder(
+    scaffoldState: ScaffoldState = rememberScaffoldState()
+): ScaffoldStateHolder {
+    val navController = rememberNavController()
+    val coroutineScope = rememberCoroutineScope()
+    val lifecycle = LocalLifecycleOwner.current.lifecycle
+    val snackbarHostState = scaffoldState.snackbarHostState
+
+    return rememberSaveable(
+        inputs = arrayOf(navController, snackbarHostState, coroutineScope, lifecycle),
+        saver = ScaffoldStateHolderSaver(
+            navController, snackbarHostState, coroutineScope, lifecycle
+        )
+    ) {
+        ScaffoldStateHolder(
+            navController, scaffoldState.snackbarHostState, coroutineScope, lifecycle, listOf()
+        )
+    }
+}
+
+/**
+ * State holder for Jetsnack's app scaffold that contains state related to [JetsnackScaffold], and
+ * handles Navigation and Snackbar events.
+ */
+class ScaffoldStateHolder(
+    val navController: NavHostController,
+    private val snackbarHostState: SnackbarHostState,
+    coroutineScope: CoroutineScope,
+    lifecycle: Lifecycle,
+    initialSnackbarMessages: List<String>
+) {
+    // Tabs for JetsnackBottomBar
+    val tabs = HomeSections.values()
+
+    // Queue a maximum of 3 snackbar messages
+    private val snackbarMessages = Channel<String>(3, BufferOverflow.DROP_OLDEST)
```

Comment:
The `BufferOverflow.DROP_OLDEST` here doesn't interact with the `pendingSnackbarMessages`.

If a message if enqueued, but is later dropped due to the `Channel` overflowing, that dropped message isn't removed from the `pendingSnackbarMessages` since it will never be received.

Reply:
I don't quite get what you mean. The message is only added to `pendingSnackbarMessages` _iff_ it could be added to the `Channel`. See L135.

Reply:
Right, the issue I'm seeing is if the `DROP_OLDEST` comes into play if there are already `3` messages in the `Channel`'s buffer.

`trySend` will always succeed (if the channel isn't closed), since it will kick out the oldest message from the `Channel`, but that removed message isn't ever removed from `pendingSnackbarMessages`.

So upon state saving/restoration, there could be really old messages that were dropped from the `Channel`, but weren't removed from `pendingSnackbarMessages`.

Reply:
Changed the code for this. Please take another look and let me know what you think! Thanks!